jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp

```diff
@@ -721,61 +749,66 @@ size_t
-  std::vector<std::vector<PointerObjectIndex>> loadConstraints(Set_.NumPointerObjects());
+  // The constraints need to be added to the unification root, as only unification roots
+  // are allowed on the worklist.
+  std::vector<util::HashSet<PointerObjectIndex>> storeConstraints(Set_.NumPointerObjects());
+  std::vector<util::HashSet<PointerObjectIndex>> loadConstraints(Set_.NumPointerObjects());
   std::vector<std::vector<const jlm::llvm::CallNode *>> callConstraints(Set_.NumPointerObjects());
 
   for (const auto & constraint : Constraints_)
   {
     if (const auto * ssConstraint = std::get_if<SupersetConstraint>(&constraint))
     {
       // Superset constraints become edges in the superset graph
-      const auto superset = ssConstraint->GetSuperset();
-      const auto subset = ssConstraint->GetSubset();
-      JLM_ASSERT(superset < Set_.NumPointerObjects() && subset < Set_.NumPointerObjects());
+      auto superset = Set_.GetUnificationRoot(ssConstraint->GetSuperset());
+      auto subset = Set_.GetUnificationRoot(ssConstraint->GetSubset());
 
       supersetEdges[subset].Insert(superset);
     }
     else if (const auto * storeConstraint = std::get_if<StoreConstraint>(&constraint))
     {
-      const auto pointer = storeConstraint->GetPointer();
-      const auto value = storeConstraint->GetValue();
-      JLM_ASSERT(pointer < Set_.NumPointerObjects() && value < Set_.NumPointerObjects());
+      auto pointer = Set_.GetUnificationRoot(storeConstraint->GetPointer());
+      auto value = Set_.GetUnificationRoot(storeConstraint->GetValue());
 
-      storeConstraints[pointer].push_back(value);
+      storeConstraints[pointer].Insert(value);
     }
     else if (const auto * loadConstraint = std::get_if<LoadConstraint>(&constraint))
     {
-      const auto pointer = loadConstraint->GetPointer();
-      const auto value = loadConstraint->GetValue();
-      JLM_ASSERT(pointer < Set_.NumPointerObjects() && value < Set_.NumPointerObjects());
+      auto pointer = Set_.GetUnificationRoot(loadConstraint->GetPointer());
+      auto value = Set_.GetUnificationRoot(loadConstraint->GetValue());
 
-      loadConstraints[pointer].push_back(value);
+      loadConstraints[pointer].Insert(value);
     }
     else if (const auto * callConstraint = std::get_if<FunctionCallConstraint>(&constraint))
     {
-      const auto pointer = callConstraint->GetPointer();
+      auto pointer = Set_.GetUnificationRoot(callConstraint->GetPointer());
       const auto & callNode = callConstraint->GetCallNode();
-      JLM_ASSERT(pointer < Set_.NumPointerObjects());
 
       callConstraints[pointer].push_back(&callNode);
     }
   }
 
-  // The worklist, initialized with every object
+  // The worklist, initialized with every unification root
   util::LrfWorklist<PointerObjectIndex> worklist;
   for (PointerObjectIndex i = 0; i < Set_.NumPointerObjects(); i++)
-    worklist.PushWorkItem(i);
+  {
+    if (Set_.GetUnificationRoot(i) == i)
```

Comment:
`Set_.IsUnificationRoot(i)` and then you implement this method with exactly the expression from above in its body.